### PR TITLE
Make ProgBarLogger, ModelCheckpoint and TensorBoard callbacks not use batch hooks when possible

### PR DESCRIPTION
**Pull request by lgeiger**
This PR changes the `ProgBarLogger`, `ModelCheckpoint` and `TensorBoard` callbacks to not use batch hooks when possible.

`ProgBarLogger` will now only require batch hooks if `verbose == 1` or the cardinality of the dataset cannot be determined in v1 execution. `ModelCheckpoint` will require batch hooks only if `save_freq != "epoch"` and `TensorBoard` only when profiling is enabled.

These changes may enable future optimizations where the entire epoch iteration in the training loop could be wrapped in a `tf.function` if no batch hooks need to be called:
https://github.com/tensorflow/tensorflow/blob/9b13f1b47fc97cb8b167d7fe5f1dc2277c85698d/tensorflow/python/keras/engine/training.py#L1089-L1103

The 2.3.0-rc0 release notes mention that `experimental_steps_per_execution` can result in up to 3x preformance improvements on TPUs. Unfortunately in my opinions the usablility of `experimental_steps_per_execution` is not really optimal since exposes too many implementation details to the user. I think it would be a lot nicer if performance optimizations like this could be done automatically for common use cases where only epoch level logging is required so users only need to care about `experimental_steps_per_execution` when requiring more customizability.
This PR is intended as a first step to explore if we can wrap the entire epoch iterations into a `tf.function` if no batch hooks need to be called, I am happy to explore this in a future PR.

This fixed `b/150629188`

### CODE REVIEW

**Comment by lgeiger** [comment]
@omalleyt12 Do you mind taking a look at this?

**Review by omalleyt12 — APPROVED** [review]
This is great! Thank you for the PR!

**Comment by lgeiger** [comment]
@omalleyt12 Thanks for merging.

It looks like wrapping an entire epoch in a `tf.function` is now possible for a few common cases where no batch level callbacks are needed.

I was playing around with something like this:
```diff
diff --git a/tensorflow/python/keras/engine/training.py b/tensorflow/python/keras/engine/training.py
index bbab6fc7f9..d1c53b6a9b 100644
--- a/tensorflow/python/keras/engine/training.py
+++ b/tensorflow/python/keras/engine/training.py
@@ -1085,25 +1085,41 @@ class Model(base_layer.Layer, version_utils.ModelVersionSelector):
       # happen after `callbacks.on_train_begin`.
       data_handler._initial_epoch = (  # pylint: disable=protected-access
           self._maybe_load_initial_epoch_from_ckpt(initial_epoch))
+
+      # maybe also try this on the second epoch when more cases will have data_handler.inferred_steps set
+      run_compiled = (not self.run_eagerly and
+          not callbacks._should_call_train_batch_hooks and
+          data_handler.inferred_steps is not None and
+          data_handler.inferred_steps > 1)
+
+      if run_compiled:
+        @def_function.function(experimental_relax_shapes=True)
+        def epoch_train_function(iterator, steps):
+          logs = train_function(iterator)
+          for _ in math_ops.range(steps - 1):
+            logs = train_function(iterator)
+          return logs
+
       logs = None
       for epoch, iterator in data_handler.enumerate_epochs():
         self.reset_metrics()
         callbacks.on_epoch_begin(epoch)
-        with data_handler.catch_stop_iteration():
-          for step in data_handler.steps():
-            with trace.Trace(
-                'train',
-                epoch_num=epoch,
-                step_num=step,
-                batch_size=batch_size,
-                _r=1):
-              callbacks.on_train_batch_begin(step)
-              tmp_logs = train_function(iterator)
-              if data_handler.should_sync:
-                context.async_wait()
-              logs = tmp_logs  # No error, now safe to assign to logs.
-              end_step = step + data_handler.step_increment
-              callbacks.on_train_batch_end(end_step, logs)
+        if run_compiled:
+          logs = epoch_train_function(iterator, data_handler.inferred_steps)
+        else:
+          with data_handler.catch_stop_iteration():
+            for step in data_handler.steps():
+              with trace.Trace(
+                  'train',
+                  epoch_num=epoch,
+                  step_num=step,
+                  batch_size=batch_size,
+                  _r=1):
+                callbacks.on_train_batch_begin(step)
+                tmp_logs = train_function(iterator)
+                if data_handler.should_sync:
+                  context.async_wait()
+                logs = tmp_logs  # No error, now safe to assign to logs.
+                end_step = step + data_handler.step_increment
+                callbacks.on_train_batch_end(end_step, logs)

         if logs is None:
           raise ValueError('Expect x to be a non-empty array or dataset.')
```
But I won't have time to flesh this our into a PR (this code sample will probably fail when used together with `experimental_steps_per_execution` and hasn't been properly benchmarked), but I thought I mention it here in case anyone wants to pick it up.

**Comment by omalleyt12** [comment]
@lgeiger Thanks! I really like this idea. It would be cool if we could default to the most performant option like this

Probably the only concern right now is that the way we compile multiple batches into one `tf.function` right now actually ends up with a 2x larger TF Graph than doing each batch separately. This has the potential to cause some issues, I think especially for TPUs. The reason is we call one batch before we enter the for-loop, so that inside the for-loop we know what the structure of the outputs is. There's some `tf.function` and `tf.data` work going on right now that should make this unnecessary in the future though in which case we should do this

**Comment by lgeiger** [comment]
@omalleyt12 Thanks for the fast answer, that makes a lot of sense.

> The reason is we call one batch before we enter the for-loop, so that inside the for-loop we know what the structure of the outputs is.

If the structure of the output is the only limitation here, I guess one could run the first epoch loop eagerly and only compile the entire loop starting from the second epoch. This would have the additional benefit that `data_handler.inferred_steps` will be defined in more cases. Although I guess having proper support for it in `tf.function` would be nicer.

I am thinking of something similar to this:
```diff
--- a/tensorflow/python/keras/engine/training.py
+++ b/tensorflow/python/keras/engine/training.py
@@ -1083,27 +1083,40 @@ class Model(base_layer.Layer, version_utils.ModelVersionSelector):
       # Handle fault-tolerance for multi-worker.
       # TODO(omalleyt): Fix the ordering issues that mean this has to
       # happen after `callbacks.on_train_begin`.
-      data_handler._initial_epoch = (  # pylint: disable=protected-access
-          self._maybe_load_initial_epoch_from_ckpt(initial_epoch))
-      logs = None
+      initial_epoch = self._maybe_load_initial_epoch_from_ckpt(initial_epoch)
+      data_handler._initial_epoch = initial_epoch  # pylint: disable=protected-access
+      logs, epoch_train_function = None, None
       for epoch, iterator in data_handler.enumerate_epochs():
         self.reset_metrics()
         callbacks.on_epoch_begin(epoch)
-        with data_handler.catch_stop_iteration():
-          for step in data_handler.steps():
-            with trace.Trace(
-                'train',
-                epoch_num=epoch,
-                step_num=step,
-                batch_size=batch_size,
-                _r=1):
-              callbacks.on_train_batch_begin(step)
-              tmp_logs = train_function(iterator)
-              if data_handler.should_sync:
-                context.async_wait()
-              logs = tmp_logs  # No error, now safe to assign to logs.
-              end_step = step + data_handler.step_increment
-              callbacks.on_train_batch_end(end_step, logs)
+        if epoch_train_function or (not self.run_eagerly and
+                                    not callbacks._should_call_train_batch_hooks
+                                    and data_handler.inferred_steps is not None
+                                    and epoch - initial_epoch > 0):
+          if not epoch_train_function:
+            @def_function.function(experimental_relax_shapes=True)
+            def epoch_train_function(iterator, steps, logs):
+              for _ in math_ops.range(steps):
+                logs = train_function(iterator)
+              return logs
+          logs = epoch_train_function(iterator, data_handler.inferred_steps, logs)
+        else:
+          with data_handler.catch_stop_iteration():
+            for step in data_handler.steps():
+              with trace.Trace(
+                  'train',
+                  epoch_num=epoch,
+                  step_num=step,
+                  batch_size=batch_size,
+                  _r=1):
+                callbacks.on_train_batch_begin(step)
+                tmp_logs = train_function(iterator)
+                if data_handler.should_sync:
+                  context.async_wait()
+                logs = tmp_logs  # No error, now safe to assign to logs.
+                end_step = step + data_handler.step_increment
+                callbacks.on_train_batch_end(end_step, logs)

         if logs is None:
           raise ValueError('Expect x to be a non-empty array or dataset.')
```

**Comment by lgeiger** [comment]
@omalleyt12 I submitted a followup PR in #41742 to take advantage of these changes in `model.fit()`
